rcon/hooks.py

```diff
@@ -359,61 +368,188 @@ def update_player_steaminfo_on_connect(rcon, struct_log, _, steam_id_64):
+                    name,
+                    steam_id_64,
+                    e,
+                )
+        elif action == RconInvalidNameActionType.ban:
+            try:
+                rcon.do_temp_ban(
+                    steam_id_64=steam_id_64,
+                    reason=config.pineapple_name_player_message,
+                    by=config.audit_message_author,
+                    duration_hours=config.ban_length_hours,
+                )
+            except Exception as e:
+                logger.error(
+                    "Could not temp ban player %s/%s: %s",
+                    name,
+                    steam_id_64,
+                    e,
+                )
+
+    # The player might not yet have finished connecting in order to action them.
+    if is_whitespace_name:
+        func = notify_whitespace_player
+    else:
+        func = notify_pineapple_player
 
-    # The player might not yet have finished connecting in order to send messages.
-    t = Timer(10, notify_player)
-    pendingTimers[steam_id_64] = t
+    t = Timer(10, func, kwargs={"action": action})
+    pendingTimers[steam_id_64].append((action, t))
     t.start()
 
+    # TODO: it is not possible to kick pineapple names, remove them by banning/unbanning
+    if is_pineappple_name and action == RconInvalidNameActionType.kick:
+        # Give the game server time to update bans so the ban can be removed automatically
+        # and give the notify timer time to finish so the player is actually banned before
+        # trying to remove it
+        # players can't connect if they're banned so this should never fire and remove a
+        # temp ban from any other reason unless there's another hook temp banning on connect
+        t = Timer(15, rcon.do_remove_temp_ban, kwargs={"steam_id_64": steam_id_64})
+        send_to_discord_audit(
+            message=config.audit_kick_unban_message, by=config.audit_message_author
+        )
+        pendingTimers[steam_id_64].append((action, t))
+        t.start()
+
 
 @on_disconnected
 @inject_player_ids
-def cleanup_pending_timers(_, _1, _2, steam_id_64: str):
-    pt: Timer = pendingTimers.pop(steam_id_64, None)
-    if pt is None:
-        return
-    if pt.is_alive():
-        try:
-            pt.cancel()
-        except:
-            pass
+def cleanup_pending_timers(rcon: Rcon, struct_log, name, steam_id_64: str):
+    """Cancel pending timers created by notify_player if the player disconnects
+
+    Only messaging the player should be cancelled if they disconnect early
+    Kicking players is non functional (RCON bug) so they're temp banned/pardoned
+    Temporary banning players doesn't create a timer
+    """
+    pts: list[tuple[RconInvalidNameActionType, Timer]] = pendingTimers.pop(
+        steam_id_64, []
+    )
+    for action, pt in pts:
+        # TODO: Can't kick pineapple names, don't cancel the thread if the ban removal is pending
+        if action != RconInvalidNameActionType.kick and pt.is_alive():
```

Comment:
this would result in timers kept, when a player disconnects on it's own with a pineapple name, right? 
However, I think this is not a big issue, the command will simply fail then which is fine.